File: apps/macos/DNSPilotMac/Sources/DNSPilotMacCore/BenchmarkProgressViewModel.swift
public enum BenchmarkFailureStep: String, Equatable, Sendable {
    case preparingBenchmark
    case resolvingDNS
    case measuringConnection
    case parsingResult
    case savingHistory

    public var label: String {
        switch self {
        case .preparingBenchmark:
            "Preparing benchmark"
        case .resolvingDNS:
            "Resolving DNS"
        case .measuringConnection:
            "Measuring TCP"
        case .parsingResult:
            "Parsing result"
        case .savingHistory:
            "Saving history"
        }
    }
}

public struct BenchmarkExecutionFailure: Equatable, Sendable, ExpressibleByStringLiteral {
    public let message: String
    public let failedStep: BenchmarkFailureStep
    public let suggestion: String
    public let debugLog: String

    public init(
        message: String,
        failedStep: BenchmarkFailureStep,
        suggestion: String? = nil,
        debugLog: String
    ) {
        self.message = message
        self.failedStep = failedStep
        self.suggestion = suggestion ?? Self.defaultSuggestion(for: failedStep)
        self.debugLog = debugLog
    }

    public init(stringLiteral value: StringLiteralType) {
        self.init(message: value, failedStep: .preparingBenchmark, debugLog: value)
    }

    public func issueReport(modeLabel: String, elapsedMS: Int?) -> String {
        var lines = [
            "Benchmark failed",
            "Mode: \(modeLabel)",
            "Failed at: \(failedStep.label)",
            "Reason: \(message)",
            "Suggestion: \(suggestion)",
        ]
        if let elapsedMS {
            lines.append("Elapsed: \(elapsedMS) ms")
        }
        lines.append("")
        lines.append("Debug log:")
        lines.append(debugLog)
        return lines.joined(separator: "\n")
    }

    private static func defaultSuggestion(for step: BenchmarkFailureStep) -> String {
        switch step {
        case .preparingBenchmark:
            "Check selected profiles, target domains, and CLI availability."
        case .resolvingDNS:
            "Try DNS + TCP or check resolver, firewall, VPN, or network configuration."
        case .measuringConnection:
            "Check network reachability, firewall, VPN, captive portal, or try DNS only."
        case .parsingResult:
            "Keep the debug log and verify the CLI output schema matches the app version."
        case .savingHistory:
            "Check Application Support storage permissions and available disk space."
        }
    }
}

public enum BenchmarkProgressStatus: String, Equatable, Sendable {
    case idle
    case running
    case success
    case degraded
    case failed
}

public struct BenchmarkProgressStepViewModel: Equatable, Identifiable, Sendable {
    public let id: String
    public let title: String
    public let status: BenchmarkProgressStatus

    public init(id: String, title: String, status: BenchmarkProgressStatus) {
        self.id = id
        self.title = title
        self.status = status
    }
}

public struct BenchmarkProgressResolverTarget: Equatable, Identifiable, Sendable {
    public let id: String
    public let name: String
    public let resolver: String

    public init(id: String, name: String, resolver: String) {
        self.id = id
        self.name = name
        self.resolver = resolver
    }
}

public struct BenchmarkResolverStatusViewModel: Equatable, Identifiable, Sendable {
    public let id: String
    public let name: String
    public let resolver: String
    public let status: BenchmarkProgressStatus
    public let detail: String

    public init(
        id: String,
        name: String,
        resolver: String,
        status: BenchmarkProgressStatus,
        detail: String
    ) {
        self.id = id
        self.name = name
        self.resolver = resolver
        self.status = status
        self.detail = detail
    }
}

public struct BenchmarkProgressPlanSummary: Equatable, Sendable {
    public let resolverCount: Int
    public let domainCount: Int
    public let attempts: Int
    public let dnsTimeoutMS: Int
    public let connectTimeoutMS: Int
    public let maxConnectTargetsPerDomain: Int
    public let recordFamily: BenchmarkRecordFamily
    public let resolverTargets: [BenchmarkProgressResolverTarget]

    public init(
        resolverCount: Int,
        domainCount: Int,
        attempts: Int,
        dnsTimeoutMS: Int = 800,
        connectTimeoutMS: Int = 1_000,
        maxConnectTargetsPerDomain: Int = 4,
        recordFamily: BenchmarkRecordFamily = .both,
        resolverTargets: [BenchmarkProgressResolverTarget] = []
    ) {
        self.resolverCount = resolverCount
        self.domainCount = domainCount
        self.attempts = attempts
        self.dnsTimeoutMS = dnsTimeoutMS
        self.connectTimeoutMS = connectTimeoutMS
        self.maxConnectTargetsPerDomain = maxConnectTargetsPerDomain
        self.recordFamily = recordFamily
        self.resolverTargets = resolverTargets
    }

    public init(plan: BenchmarkPlanViewModel) {
        self.init(
            resolverCount: plan.resolverCount,
            domainCount: plan.domains.count,
            attempts: plan.attempts,
            dnsTimeoutMS: plan.dnsTimeoutMS,
            connectTimeoutMS: plan.connectTimeoutMS,
            maxConnectTargetsPerDomain: plan.maxConnectTargetsPerDomain,
            recordFamily: plan.recordFamily,
            resolverTargets: plan.resolverTargets
        )
    }
}

public struct BenchmarkProgressViewModel: Equatable, Sendable {
    public let steps: [BenchmarkProgressStepViewModel]
    public let currentStepVerboseLines: [String]
    public let resolverStatuses: [BenchmarkResolverStatusViewModel]

    public init(
        mode: BenchmarkPlanMode,
        state: BenchmarkRunState,
        outcome: BenchmarkExecutionOutcome?,
        historySaved: Bool,
        planSummary: BenchmarkProgressPlanSummary? = nil,
        progressEvents: [BenchmarkProgressEvent] = []
    ) {
        let failure: BenchmarkExecutionFailure? = {
            if case .failed(let failure) = outcome {
                return failure
            }
            return nil
        }()
        let isCancelling = {
            if case .cancelling = state {
                return true
            }
            return false
        }()
        let isActive = {
            if case .running = state {
                return true
            }
            return isCancelling
        }()
        let isCompleted = state == .completed

        var nextSteps: [BenchmarkProgressStepViewModel] = [
            Self.step(.preparingBenchmark, status: Self.status(for: .preparingBenchmark, failure: failure, isRunning: isActive, isCompleted: isCompleted)),
            Self.step(.resolvingDNS, status: Self.status(for: .resolvingDNS, failure: failure, isRunning: isActive, isCompleted: isCompleted)),
        ]

        if mode == .connectionPathCompare {
            nextSteps.append(
                Self.step(.measuringConnection, status: Self.status(for: .measuringConnection, failure: failure, isRunning: isActive, isCompleted: isCompleted))
            )
        }

        nextSteps.append(
            Self.step(.parsingResult, status: Self.status(for: .parsingResult, failure: failure, isRunning: false, isCompleted: isCompleted))
        )
        nextSteps.append(
            Self.step(.savingHistory, status: Self.historyStatus(failure: failure, isCompleted: isCompleted, historySaved: historySaved))
        )
        steps = nextSteps
        currentStepVerboseLines = Self.verboseLines(
            mode: mode,
            isRunning: isActive,
            isCancelling: isCancelling,
            planSummary: planSummary,
            progressEvents: progressEvents
        )
        resolverStatuses = Self.resolverStatuses(
            isRunning: isActive,
            isCancelling: isCancelling,
            failure: failure,
            outcome: outcome,
            targets: planSummary?.resolverTargets ?? [],
            progressEvents: progressEvents
        )
    }

    private static func step(
        _ failureStep: BenchmarkFailureStep,
        status: BenchmarkProgressStatus
    ) -> BenchmarkProgressStepViewModel {
        BenchmarkProgressStepViewModel(
            id: failureStep.rawValue,
            title: failureStep.label,
            status: status
        )
    }

    private static func status(
        for step: BenchmarkFailureStep,
        failure: BenchmarkExecutionFailure?,
        isRunning: Bool,
        isCompleted: Bool
    ) -> BenchmarkProgressStatus {
        if let failure {
            if failure.failedStep == step {
                return .failed
            }
            return order(step) < order(failure.failedStep) ? .success : .idle
        }
        if isCompleted {
            return .success
        }
        if isRunning {
            switch step {
            case .preparingBenchmark:
                return .success
            case .resolvingDNS, .measuringConnection:
                return .running
            case .parsingResult, .savingHistory:
                return .idle
            }
        }
        return .idle
    }

    private static func historyStatus(
        failure: BenchmarkExecutionFailure?,
        isCompleted: Bool,
        historySaved: Bool
    ) -> BenchmarkProgressStatus {
        if failure?.failedStep == .savingHistory {
            return .failed
        }
        return isCompleted && historySaved ? .success : .idle
    }

    private static func order(_ step: BenchmarkFailureStep) -> Int {
        switch step {
        case .preparingBenchmark:
            0
        case .resolvingDNS:
            1
        case .measuringConnection:
            2
        case .parsingResult:
            3
        case .savingHistory:
            4
        }
    }

    private static func verboseLines(
        mode: BenchmarkPlanMode,
        isRunning: Bool,
        isCancelling: Bool,
        planSummary: BenchmarkProgressPlanSummary?,
        progressEvents: [BenchmarkProgressEvent]
    ) -> [String] {
        guard isRunning else {
            return []
        }
        if isCancelling {
            return [
                "* Cancellation requested; waiting for the CLI process to stop.",
                "* Output is still drained so the final state and debug log stay consistent.",
            ]
        }
        guard let planSummary else {
            return [
                "* Running benchmark command.",
                "* Waiting for CLI output; cancel is available if the network blocks.",
                "* CLI probes resolvers sequentially; per-resolver rows update from progress events when available.",
            ]
        }
        if let progressLines = progressEventVerboseLines(
            progressEvents,
            planSummary: planSummary
        ) {
            return progressLines
        }

        let dnsSeconds = worstCaseDNSSeconds(summary: planSummary)
        let tcpSeconds = worstCaseTCPSeconds(summary: planSummary)
        switch mode {
        case .dnsOnlyCompare:
            return [
                "* Resolving \(planSummary.domainCount) domain(s) with \(planSummary.resolverCount) resolver(s), \(planSummary.attempts) attempt(s), \(planSummary.recordFamily.displayLabel).",
                "* Worst-case DNS wait before output: about \(dnsSeconds); stdout is drained while the CLI runs.",
                "* CLI probes resolvers sequentially; per-resolver rows update from progress events when available.",
            ]
        case .connectionPathCompare:
            return [
                "* Resolving DNS, then probing TCP :443 for returned endpoints.",
                "* Planned input: \(planSummary.domainCount) domain(s), \(planSummary.resolverCount) resolver(s), \(planSummary.attempts) attempt(s); worst-case DNS phase about \(dnsSeconds), TCP phase about \(tcpSeconds).",
                "* CLI probes resolvers sequentially; per-resolver rows update from progress events when available.",
            ]
        }
    }

    private static func progressEventVerboseLines(
        _ progressEvents: [BenchmarkProgressEvent],
        planSummary: BenchmarkProgressPlanSummary
    ) -> [String]? {
        guard let event = progressEvents.last else {
            return nil
        }
        let resolverName = planSummary.resolverTargets.first { $0.id == event.profileID }?.name ?? event.profileID
        switch event.type {
        case .resolverStarted:
            return [
                "* Current resolver: \(resolverName) (\(event.resolver)), \(event.index)/\(event.total).",
                "* Waiting for this resolver to finish; elapsed time is shown on completion.",
            ]
        case .resolverFinished:
            return [
                "* Last finished: \(resolverName) (\(event.resolver)), \(detail(for: event)).",
                "* Waiting for the next resolver event or final JSON output.",
            ]
        }
    }

    private static func resolverStatuses(
        isRunning: Bool,
        isCancelling: Bool,
        failure: BenchmarkExecutionFailure?,
        outcome: BenchmarkExecutionOutcome?,
        targets: [BenchmarkProgressResolverTarget],
        progressEvents: [BenchmarkProgressEvent]
    ) -> [BenchmarkResolverStatusViewModel] {
        if case .completed(let resultViewModel) = outcome {
            return resultViewModel.rows.map { row in
                BenchmarkResolverStatusViewModel(
                    id: row.profileID,
                    name: row.name,
                    resolver: row.resolver,
                    status: row.status,
                    detail: row.statusDetail
                )
            }
        }

        if isRunning {
            if !progressEvents.isEmpty {
                let latestEvents = latestProgressEventsByProfileID(progressEvents)
                return targets.map { target in
                    if let event = latestEvents[target.id] {
                        return BenchmarkResolverStatusViewModel(
                            id: target.id,
                            name: target.name,
                            resolver: target.resolver,
                            status: status(for: event),
                            detail: detail(for: event)
                        )
                    }

                    return BenchmarkResolverStatusViewModel(
                        id: target.id,
                        name: target.name,
                        resolver: target.resolver,
                        status: .idle,
                        detail: "Pending"
                    )
                }
            }

            return targets.map { target in
                BenchmarkResolverStatusViewModel(
                    id: target.id,
                    name: target.name,
                    resolver: target.resolver,
                    status: .running,
                    detail: isCancelling ? "Cancelling" : "Waiting for final JSON"
                )
            }
        }

        if let failure {
            return targets.map { target in
                BenchmarkResolverStatusViewModel(
                    id: target.id,
                    name: target.name,
                    resolver: target.resolver,
                    status: failure.failedStep == .preparingBenchmark ? .idle : .failed,
                    detail: failure.failedStep == .parsingResult ? "Result parsing failed" : "Benchmark failed"
                )
            }
        }

        return []
    }

    private static func latestProgressEventsByProfileID(
        _ events: [BenchmarkProgressEvent]
    ) -> [String: BenchmarkProgressEvent] {
        var latestEvents: [String: BenchmarkProgressEvent] = [:]
        for event in events {
            latestEvents[event.profileID] = event
        }
        return latestEvents
    }

    private static func status(for event: BenchmarkProgressEvent) -> BenchmarkProgressStatus {
        switch event.type {
        case .resolverStarted:
            return .running
        case .resolverFinished:
            switch event.status {
            case .success:
                return .success
            case .degraded:
                return .degraded
            case .failed:
                return .failed
            case nil:
                return .success
            }
        }
    }

    private static func detail(for event: BenchmarkProgressEvent) -> String {
        switch event.type {
        case .resolverStarted:
            return "Running \(event.index)/\(event.total)"
        case .resolverFinished:
            let summary = event.failureRate.map { "\(percent($0))% failed" } ?? "Finished"
            guard let elapsedMS = event.elapsedMS else {
                return summary
            }
            return "\(summary) - \(formatElapsedMS(elapsedMS))"
        }
    }

    private static func percent(_ value: Double) -> Int {
        Int((min(max(value, 0), 1) * 100).rounded())
    }

    private static func formatElapsedMS(_ value: Double) -> String {
        "\(Int(max(value, 0).rounded())) ms"
    }

    private static func worstCaseDNSSeconds(summary: BenchmarkProgressPlanSummary) -> String {
        let totalMilliseconds = summary.resolverCount
            * summary.domainCount
            * summary.recordFamily.recordTypeCount
            * summary.attempts
            * summary.dnsTimeoutMS
        let seconds = Double(totalMilliseconds) / 1_000
        return String(format: "%.1fs", seconds)
    }

    private static func worstCaseTCPSeconds(summary: BenchmarkProgressPlanSummary) -> String {
        let totalMilliseconds = summary.resolverCount
            * summary.domainCount
            * summary.maxConnectTargetsPerDomain
            * summary.attempts
            * summary.connectTimeoutMS
        let seconds = Double(totalMilliseconds) / 1_000
        return String(format: "%.1fs", seconds)
    }
}

File: apps/windows/DNSPilotWindows/src/DNSPilotWindows.Core/BenchmarkProgressViewModel.cs
namespace DNSPilotWindows.Core;

public enum BenchmarkRunState
{
    Idle,
    Running,
    Cancelling,
    Completed,
}

public enum ProgressStatus
{
    Idle,
    Running,
    Success,
    Degraded,
    Failed,
}

public readonly record struct BenchmarkFailureStep(string Id, string Label, int Order)
{
    public static BenchmarkFailureStep PreparingBenchmark { get; } = new("preparingBenchmark", "Preparing benchmark", 0);
    public static BenchmarkFailureStep ResolvingDns { get; } = new("resolvingDNS", "Resolving DNS", 1);
    public static BenchmarkFailureStep MeasuringConnection { get; } = new("measuringConnection", "Measuring TCP", 2);
    public static BenchmarkFailureStep ParsingResult { get; } = new("parsingResult", "Parsing result", 3);
    public static BenchmarkFailureStep SavingHistory { get; } = new("savingHistory", "Saving history", 4);
}

public sealed record BenchmarkProgressStepViewModel(
    string Id,
    string Title,
    ProgressStatus Status);

public sealed record BenchmarkResolverStatusViewModel(
    string Id,
    string Name,
    string Resolver,
    ProgressStatus Status,
    string Detail);

public enum ProgressEventType
{
    ResolverStarted,
    ResolverFinished,
}

public enum ProgressEventStatus
{
    Success,
    Degraded,
    Failed,
}

public sealed record BenchmarkProgressEvent(
    ProgressEventType Type,
    string ProfileId,
    string Resolver,
    int Index,
    int Total,
    ProgressEventStatus? Status = null,
    double? FailureRate = null,
    double? ElapsedMs = null);

public sealed class BenchmarkExecutionFailure
{
    public BenchmarkExecutionFailure(
        BenchmarkFailureStep failedStep,
        string reason,
        TimeSpan? elapsed,
        string debugLog,
        string? suggestion = null)
    {
        FailedStep = failedStep;
        Reason = reason;
        Elapsed = elapsed;
        DebugLog = debugLog;
        Suggestion = suggestion;
    }

    public BenchmarkFailureStep FailedStep { get; }
    public string Reason { get; }
    public TimeSpan? Elapsed { get; }
    public string DebugLog { get; }
    public string? Suggestion { get; }

    public string CopyableReport(string modeLabel)
    {
        var lines = new List<string>
        {
            WindowsDisplayText.Text("Benchmark failed", "Benchmark thất bại"),
            $"{WindowsDisplayText.Text("Mode", "Chế độ")}: {modeLabel}",
            $"{WindowsDisplayText.Text("Failed at", "Bước lỗi")}: {WindowsDisplayText.StepLabel(FailedStep)}",
            $"{WindowsDisplayText.Text("Reason", "Lý do")}: {Reason}",
            $"{WindowsDisplayText.Text("Suggestion", "Gợi ý")}: {Suggestion ?? DefaultSuggestion(FailedStep)}",
        };

        if (Elapsed is { } elapsed)
        {
            lines.Add($"{WindowsDisplayText.Text("Elapsed", "Thời gian")}: {Math.Round(elapsed.TotalMilliseconds):0} ms");
        }

        lines.Add(string.Empty);
        lines.Add(WindowsDisplayText.Text("Debug log:", "Log debug:"));
        lines.Add(DebugLog);
        return string.Join(Environment.NewLine, lines);
    }

    private static string DefaultSuggestion(BenchmarkFailureStep step)
    {
        if (step == BenchmarkFailureStep.MeasuringConnection)
        {
            return WindowsDisplayText.Text(
                "Check network reachability, firewall, VPN, captive portal, or try DNS only.",
                "Kiểm tra kết nối mạng, firewall, VPN, captive portal, hoặc thử chế độ chỉ DNS.");
        }

        if (step == BenchmarkFailureStep.ParsingResult)
        {
            return WindowsDisplayText.Text(
                "Keep the debug log and verify the CLI output schema matches the app version.",
                "Giữ log debug và xác minh schema output của CLI khớp phiên bản app.");
        }

        if (step == BenchmarkFailureStep.SavingHistory)
        {
            return WindowsDisplayText.Text(
                "Check local DNS Pilot storage permissions and available disk space.",
                "Kiểm tra quyền lưu trữ cục bộ của DNS Pilot và dung lượng đĩa.");
        }

        return WindowsDisplayText.Text(
            "Check selected profiles, target domains, CLI availability, and network configuration.",
            "Kiểm tra hồ sơ đã chọn, domain test, CLI, và cấu hình mạng.");
    }
}

public sealed class BenchmarkProgressViewModel
{
    private BenchmarkProgressViewModel(
        IReadOnlyList<BenchmarkProgressStepViewModel> steps,
        IReadOnlyList<string> currentStepLines,
        IReadOnlyList<BenchmarkResolverStatusViewModel> resolverStatuses)
    {
        Steps = steps;
        CurrentStepLines = currentStepLines;
        ResolverStatuses = resolverStatuses;
    }

    public IReadOnlyList<BenchmarkProgressStepViewModel> Steps { get; }
    public IReadOnlyList<string> CurrentStepLines { get; }
    public IReadOnlyList<BenchmarkResolverStatusViewModel> ResolverStatuses { get; }

    public static BenchmarkProgressViewModel From(
        BenchmarkMode mode,
        BenchmarkRunState state,
        BenchmarkProgressPlanSummary summary,
        IReadOnlyList<BenchmarkProgressEvent>? progressEvents = null,
        BenchmarkExecutionFailure? failure = null,
        bool historySaved = false)
    {
        progressEvents ??= Array.Empty<BenchmarkProgressEvent>();
        var running = state is BenchmarkRunState.Running or BenchmarkRunState.Cancelling;
        var completed = state == BenchmarkRunState.Completed && failure is null;
        var steps = StepSequence(mode)
            .Select(step => new BenchmarkProgressStepViewModel(step.Id, WindowsDisplayText.StepLabel(step), StepStatus(step, failure, running, completed, historySaved)))
            .ToArray();

        return new BenchmarkProgressViewModel(
            steps,
            BuildCurrentStepLines(mode, running, state == BenchmarkRunState.Cancelling, summary, progressEvents),
            BuildResolverStatuses(running, state == BenchmarkRunState.Cancelling, summary, progressEvents, failure));
    }

    private static IReadOnlyList<BenchmarkFailureStep> StepSequence(BenchmarkMode mode)
    {
        var steps = new List<BenchmarkFailureStep>
        {
            BenchmarkFailureStep.PreparingBenchmark,
            BenchmarkFailureStep.ResolvingDns,
        };

        if (mode == BenchmarkMode.DnsAndTcp)
        {
            steps.Add(BenchmarkFailureStep.MeasuringConnection);
        }

        steps.Add(BenchmarkFailureStep.ParsingResult);
        steps.Add(BenchmarkFailureStep.SavingHistory);
        return steps;
    }

    private static ProgressStatus StepStatus(
        BenchmarkFailureStep step,
        BenchmarkExecutionFailure? failure,
        bool running,
        bool completed,
        bool historySaved)
    {
        if (failure is not null)
        {
            if (failure.FailedStep == step)
            {
                return ProgressStatus.Failed;
            }

            return step.Order < failure.FailedStep.Order ? ProgressStatus.Success : ProgressStatus.Idle;
        }

        if (step == BenchmarkFailureStep.SavingHistory)
        {
            return completed && historySaved ? ProgressStatus.Success : ProgressStatus.Idle;
        }

        if (completed)
        {
            return ProgressStatus.Success;
        }

        if (!running)
        {
            return ProgressStatus.Idle;
        }

        if (step == BenchmarkFailureStep.PreparingBenchmark)
        {
            return ProgressStatus.Success;
        }

        return step == BenchmarkFailureStep.ResolvingDns || step == BenchmarkFailureStep.MeasuringConnection
            ? ProgressStatus.Running
            : ProgressStatus.Idle;
    }

    private static IReadOnlyList<string> BuildCurrentStepLines(
        BenchmarkMode mode,
        bool running,
        bool cancelling,
        BenchmarkProgressPlanSummary summary,
        IReadOnlyList<BenchmarkProgressEvent> progressEvents)
    {
        if (!running)
        {
            return Array.Empty<string>();
        }

        if (cancelling)
        {
            return new[]
            {
                WindowsDisplayText.Text(
                    "Cancellation requested; waiting for the CLI process to stop.",
                    "Đã yêu cầu hủy; đang chờ tiến trình CLI dừng."),
                WindowsDisplayText.Text(
                    "Output is still drained so the final state and debug log stay consistent.",
                    "Output vẫn được đọc hết để trạng thái cuối và log debug nhất quán."),
            };
        }

        if (progressEvents.LastOrDefault() is { } latest)
        {
            return new[]
            {
                WindowsDisplayText.Text(
                    $"Current resolver event: {latest.ProfileId} ({latest.Resolver}), {latest.Index}/{latest.Total}.",
                    $"Sự kiện resolver hiện tại: {latest.ProfileId} ({latest.Resolver}), {latest.Index}/{latest.Total}."),
                latest.Type == ProgressEventType.ResolverFinished
                    ? WindowsDisplayText.Text($"Last result: {EventDetail(latest)}.", $"Kết quả cuối: {EventDetail(latest)}.")
                    : WindowsDisplayText.Text("Waiting for this resolver to finish.", "Đang chờ resolver này hoàn tất."),
            };
        }

        return mode == BenchmarkMode.DnsAndTcp
            ? new[]
            {
                WindowsDisplayText.Text(
                    "Resolving DNS, then probing TCP :443 for returned endpoints.",
                    "Đang phân giải DNS, rồi kiểm tra TCP :443 cho endpoint trả về."),
                WindowsDisplayText.Text(
                    $"Planned input: {summary.DomainCount} domain(s), {summary.ResolverCount} resolver(s), {summary.Attempts} attempt(s).",
                    $"Đầu vào dự kiến: {summary.DomainCount} domain, {summary.ResolverCount} resolver, {summary.Attempts} lần thử."),
            }
            : new[]
            {
                WindowsDisplayText.Text(
                    $"Resolving {summary.DomainCount} domain(s) with {summary.ResolverCount} resolver(s), {summary.Attempts} attempt(s), {summary.RecordFamily.DisplayLabel}.",
                    $"Đang phân giải {summary.DomainCount} domain với {summary.ResolverCount} resolver, {summary.Attempts} lần thử, {WindowsDisplayText.RecordFamilyLabel(summary.RecordFamily)}."),
                WindowsDisplayText.Text(
                    "CLI probes resolvers sequentially; per-resolver rows update from progress events when available.",
                    "CLI kiểm tra resolver tuần tự; từng dòng resolver sẽ cập nhật khi có progress event."),
            };
    }

    private static IReadOnlyList<BenchmarkResolverStatusViewModel> BuildResolverStatuses(
        bool running,
        bool cancelling,
        BenchmarkProgressPlanSummary summary,
        IReadOnlyList<BenchmarkProgressEvent> progressEvents,
        BenchmarkExecutionFailure? failure)
    {
        if (running)
        {
            var latestEventsByProfile = progressEvents
                .GroupBy(progressEvent => progressEvent.ProfileId)
                .ToDictionary(group => group.Key, group => group.Last(), StringComparer.Ordinal);

            if (latestEventsByProfile.Count > 0)
            {
                return summary.ResolverTargets
                    .Select(target => latestEventsByProfile.TryGetValue(target.Id, out var latest)
                        ? new BenchmarkResolverStatusViewModel(target.Id, target.Name, target.Resolver, EventStatus(latest), EventDetail(latest))
                        : new BenchmarkResolverStatusViewModel(target.Id, target.Name, target.Resolver, ProgressStatus.Idle, WindowsDisplayText.Text("Pending", "Đang chờ")))
                    .ToArray();
            }

            return summary.ResolverTargets
                .Select((target, index) => new BenchmarkResolverStatusViewModel(
                    target.Id,
                    target.Name,
                    target.Resolver,
                    index == 0 ? ProgressStatus.Running : ProgressStatus.Idle,
                    cancelling && index == 0
                        ? WindowsDisplayText.Text("Cancelling", "Đang hủy")
                        : index == 0
                            ? WindowsDisplayText.Text($"Running 1/{summary.ResolverTargets.Count}", $"Đang chạy 1/{summary.ResolverTargets.Count}")
                            : WindowsDisplayText.Text("Pending", "Đang chờ")))
                .ToArray();
        }

        if (failure is not null)
        {
            return summary.ResolverTargets
                .Select(target => new BenchmarkResolverStatusViewModel(
                    target.Id,
                    target.Name,
                    target.Resolver,
                    failure.FailedStep == BenchmarkFailureStep.PreparingBenchmark ? ProgressStatus.Idle : ProgressStatus.Failed,
                    failure.FailedStep == BenchmarkFailureStep.ParsingResult
                        ? WindowsDisplayText.Text("Result parsing failed", "Đọc kết quả thất bại")
                        : WindowsDisplayText.Text("Benchmark failed", "Benchmark thất bại")))
                .ToArray();
        }

        return Array.Empty<BenchmarkResolverStatusViewModel>();
    }

    private static ProgressStatus EventStatus(BenchmarkProgressEvent progressEvent)
    {
        if (progressEvent.Type == ProgressEventType.ResolverStarted)
        {
            return ProgressStatus.Running;
        }

        return progressEvent.Status switch
        {
            ProgressEventStatus.Degraded => ProgressStatus.Degraded,
            ProgressEventStatus.Failed => ProgressStatus.Failed,
            _ => ProgressStatus.Success,
        };
    }

    private static string EventDetail(BenchmarkProgressEvent progressEvent)
    {
        if (progressEvent.Type == ProgressEventType.ResolverStarted)
        {
            return WindowsDisplayText.Text(
                $"Running {progressEvent.Index}/{progressEvent.Total}",
                $"Đang chạy {progressEvent.Index}/{progressEvent.Total}");
        }

        var summary = progressEvent.FailureRate is { } failureRate
            ? WindowsDisplayText.Text(
                $"{Math.Round(Math.Clamp(failureRate, 0, 1) * 100):0}% failed",
                $"{Math.Round(Math.Clamp(failureRate, 0, 1) * 100):0}% lỗi")
            : WindowsDisplayText.Text("Finished", "Hoàn tất");

        return progressEvent.ElapsedMs is { } elapsedMs
            ? $"{summary} - {Math.Round(Math.Max(elapsedMs, 0)):0} ms"
            : summary;
    }
}
